README.md
<p align="center">
  <img src="https://media.giphy.com/media/XeGBgGGTVk514aGI0D/giphy.gif" />
  <h1 align="center" style="font-size: 48px;">🔗 Flutter Parsed text</h1>
  <h5 align="center">
A Flutter package to parse text and extract parts using predefined types like <code>url</code>, <code>phone</code> and <code>email</code> and also supports <code>Regex</code>.</h5>
</p>


## Usage 💻

To use this package, add `flutter_parsed_text` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

```dart
import 'package:flutter_parsed_text/flutter_parsed_text.dart';
```

## Working ⚙️

ParsedText can receive this paramters & all the RichText parameters:

`text`: Text that will be parsed and rendered.

`style`: It takes a `TextStyle` object as it's property to style all the non links text objects.

`parse`: Array of `MatchText` object - used for defining structure for pattern matching .

```dart
MatchText(
  type: ParsedType.EMAIL, // predefined type can be any of this ParsedTypes
  style: TextStyle(
    color: Colors.red,
    fontSize: 24,
  ), // custom style to be applied to this matched text
  onTap: (url) {
    // do something here with passed url
  }, // callback funtion when the text is tapped on
),
```

>You can also define a custom pattern like this:

```dart
MatchText(
  pattern: r"\B#+([\w]+)\b", // a custom pattern to match
  style: TextStyle(
    color: Colors.pink,
    fontSize: 24,
  ), // custom style to be applied to this matched text
  onTap: (url) async {
  // do something here with passed url
  }, // callback funtion when the text is tapped on
)
```

>You can also set RegexOption for the custom regex pattern like so:

```dart
MatchText(
  pattern: r"\B#+([\w]+)\b", // a custom pattern to match
  regexOptions: RegexOptions(
    multiLine : false,
    caseSensitive : false,
    unicode : false,
    dotAll : false
  ),
  style: TextStyle(
    color: Colors.pink,
    fontSize: 24,
  ), // custom style to be applied to this matched text
  onTap: (url) async {
  // do something here with passed url
  }, // callback funtion when the text is tapped on
)
```

A boolean that show a diffrent text and passes a diffrent text to the callback

eg: Your str is `"Mention [@michel:5455345]"` where `5455345` is ID of this user which will be passed as parameter to the callback funtion and `@michel` the value to display on interface. Your pattern for ID & username extraction : `/\[(@[^:]+):([^\]]+)\]/i`

>Displayed text will be : `Mention ^^@michel^^`

```dart
MatchText(
  pattern: r"\[(@[^:]+):([^\]]+)\]",
  style: TextStyle(
    color: Colors.green,
    fontSize: 24,
  ),
  // you must return a map with two keys
  // [display] - the text you want to show to the user
  // [value] - the value underneath it
  renderText: ({String str, String pattern}) {
    Map<String, String> map = Map<String, String>();
    RegExp customRegExp = RegExp(pattern);
    Match match = customRegExp.firstMatch(str);
    map['display'] = match.group(1);
    map['value'] = match.group(2);
    return map;
  },
  onTap: (url) {
    // do something here with passed url
  },
),
```

## Example ✍🏻

Find the complete example wiring in the [Flutter_Parsed_Text example application](https://github.com/fayeed/flutter_parsed_text/blob/master/example/lib/main.dart).

A small example of the ParsedText widget.

```dart
ParsedText(
  text:
    "[@michael:51515151] Hello this is an example of the ParsedText, links like http://www.google.com or http://www.facebook.com are clickable and phone number 444-555-6666 can call too. But you can also do more with this package, for example Bob will change style and David too. foo@gmail.com And the magic number is 42! #react #react-native",
  parse: <MatchText>[
    MatchText(
      type: ParsedType.EMAIL,
      style: TextStyle(
        color: Colors.red,
        fontSize: 24,
      ),
      onTap: (url) {
        launch("mailto:" + url);
      },
    ),
  ],
)
```

## Found this project useful? ❤️
If you found this project useful, then please consider giving it a ⭐️ on Github and sharing it with your friends via social media.

## API details 👨🏻‍💻

See the [flutter_parsed_text.dart](https://github.com/fayeed/flutter_parsed_text/blob/master/lib/flutter_parsed_text.dart) for more API details

## License ⚖️
- [MIT](https://github.com/fayeed/dash_chat/blob/master/LICENSE)

## Issues and feedback 💭

If you have any suggestion for including a feature or if something doesn't work, feel free to open a Github [issue](https://github.com/fayeed/flutter_parsed_text/issues) for us to have a discussion on it.


CHANGELOG.md
## [2.2.1]

- Fixes the issue where having an empty list of parse (MatchText) or not being able to find the mapping for the pattern lead the widget to crash.

## [2.2.0]

- Removed regexOptions property from MatchText and now comes with a global regexOptions property for ParsedText Widget.
- Changed the internals of the ParsedText widget, it now works similar to [flutter_mentions](https://github.com/fayeed/flutter_mentions)
- This also fixes issue [#11](https://github.com/fayeed/flutter_parsed_text/issues/11)

## [2.1.1]

- Add dash to the ParsedType.URL match

## [2.1.0]

- Added support for Inline widgets. Thanks to - [@angelosilvestre](https://github.com/angelosilvestre)

## [2.0.0]

- Migrates to dart null-safety. Thanks to - [@josh-burton](https://github.com/josh-burton)

## [1.2.5]

- Fixed an issue where not passing `onTap` for MatchText would result in an error.

## [1.2.4]

- Add option to make text selectable

- Add onTap callback

## [1.2.3]

- ParsedType enum added and fixed a issue where a space was added after the parsed text.

## [1.2.2]

- Regex Options added as parameter

## [1.2.1]

- Empty List as default value for parse

## [1.2.0]

- Key added & In build types inproved & Changed how text parsing works

## [1.1.2]

- README.md updated

## [1.1.1]

- README.md updated

## [1.1.0]

- Replace Text Widget with RichText Widget

## [1.0.0]

- Initial Release


lib/src/match_text.dart
part of flutter_parsed_text;

enum ParsedType { EMAIL, PHONE, URL, CUSTOM }

/// A MatchText class which provides a structure for [ParsedText] to handle
/// Pattern matching and also to provide custom [Function] and custom [TextStyle].
class MatchText {
  /// Used to enforce Predefined regex to match from
  ParsedType type;

  /// If no [type] property is explicitly defined then this propery must be
  /// non null takes a [regex] string
  String? pattern;

  /// Takes a custom style of [TextStyle] for the matched text widget
  TextStyle? style;

  /// A custom [Function] to handle onTap.
  Function(String)? onTap;

  /// A custom [Function] to handle onLongTap.
  Function(String)? onLongTap;

  /// A callback function that takes two parameter String & pattern
  ///
  /// @param str - is the word that is being matched
  /// @param pattern - pattern passed to the MatchText class
  ///
  /// eg: Your str is 'Mention [@michel:5455345]' where 5455345 is ID of this user
  /// and @michel the value to display on interface.
  /// Your pattern for ID & username extraction : `/\[(@[^:]+):([^\]]+)\]/`i
  /// Displayed text will be : Mention `@michel`
  Map<String, String> Function({
    required String str,
    required String pattern,
  })? renderText;

  /// A callback function that takes the [text] the matches the [pattern] and returns
  /// the [Widget] to be displayed inside a [WidgetSpan]
  Widget Function({
    required String text,
    required String pattern,
  })? renderWidget;

  final RegexOptions regexOptions;

  /// Creates a MatchText object
  MatchText({
    this.type = ParsedType.CUSTOM,
    this.pattern,
    this.style,
    this.onTap,
    this.onLongTap,
    this.renderText,
    this.renderWidget,
    this.regexOptions = const RegexOptions(),
  });
}


lib/src/parsed_text.dart
part of flutter_parsed_text;

/// Parse text and make them into multiple Flutter Text widgets
class ParsedText extends StatelessWidget {
  /// If non-null, the style to use for the global text.
  ///
  /// It takes a [TextStyle] object as it's property to style all the non links text objects.
  final TextStyle? style;

  /// Takes a list of [MatchText] object.
  ///
  /// This list is used to find patterns in the String and assign onTap [Function] when its
  /// tapped and also to provide custom styling to the linkify text
  final List<MatchText> parse;

  /// Text that is rendered
  ///
  /// Takes a [String]
  final String text;

  /// A text alignment property used to align the the text enclosed
  ///
  /// Uses a [TextAlign] object and default value is [TextAlign.start]
  final TextAlign alignment;

  /// A text alignment property used to align the the text enclosed
  ///
  /// Uses a [TextDirection] object and default value is [TextDirection.start]
  final TextDirection? textDirection;

  /// Whether the text should break at soft line breaks.
  ///
  ///If false, the glyphs in the text will be positioned as if there was unlimited horizontal space.
  final bool softWrap;

  /// How visual overflow should be handled.
  final TextOverflow overflow;

  /// The number of font pixels for each logical pixel.
  ///
  /// For example, if the text scale factor is 1.5, text will be 50% larger than
  /// the specified font size.
  final double textScaleFactor;

  /// An optional maximum number of lines for the text to span, wrapping if necessary.
  /// If the text exceeds the given number of lines, it will be truncated according
  /// to [overflow].
  ///
  /// If this is 1, text will not wrap. Otherwise, text will be wrapped at the
  /// edge of the box.
  final int? maxLines;

  /// {@macro flutter.painting.textPainter.strutStyle}
  final StrutStyle? strutStyle;

  /// {@macro flutter.widgets.text.DefaultTextStyle.textWidthBasis}
  final TextWidthBasis textWidthBasis;

  /// Make this text selectable.
  ///
  /// SelectableText does not support softwrap, overflow, textScaleFactor
  final bool selectable;

  /// onTap function for the whole widget
  final Function? onTap;

  /// Global regex options for the whole string,
  ///
  /// Note: Removed support for regexOptions for MatchText and now it uses global regex options.
  final RegexOptions regexOptions;

  /// Creates a parsedText widget
  ///
  /// [text] paramtere should not be null and is always required.
  /// If the [style] argument is null, the text will use the style from the
  /// closest enclosing [DefaultTextStyle].
  ParsedText({
    Key? key,
    required this.text,
    this.parse = const <MatchText>[],
    this.style,
    this.alignment = TextAlign.start,
    this.textDirection,
    this.softWrap = true,
    this.overflow = TextOverflow.clip,
    this.textScaleFactor = 1.0,
    this.strutStyle,
    this.textWidthBasis = TextWidthBasis.parent,
    this.maxLines,
    this.onTap,
    this.selectable = false,
    this.regexOptions = const RegexOptions(),
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    // Seperate each word and create a new Array
    String newString = text;

    Map<String, MatchText> _mapping = Map<String, MatchText>();

    parse.forEach((e) {
      if (e.type == ParsedType.EMAIL) {
        _mapping[emailPattern] = e;
      } else if (e.type == ParsedType.PHONE) {
        _mapping[phonePattern] = e;
      } else if (e.type == ParsedType.URL) {
        _mapping[urlPattern] = e;
      } else {
        _mapping[e.pattern!] = e;
      }
    });

    final pattern = '(${_mapping.keys.toList().join('|')})';

    List<InlineSpan> widgets = [];

    newString.splitMapJoin(
        RegExp(
          pattern,
          multiLine: regexOptions.multiLine,
          caseSensitive: regexOptions.caseSensitive,
          dotAll: regexOptions.dotAll,
          unicode: regexOptions.unicode,
        ), onMatch: (Match match) {
      final matchText = match[0];

      final mapping = _mapping[matchText!] ??
          _mapping[_mapping.keys.firstWhere((element) {
            final reg = RegExp(element);
            return reg.hasMatch(matchText);
          })]!;

      InlineSpan widget;

      if (mapping.renderText != null) {
        Map<String, String> result =
            mapping.renderText!(str: matchText, pattern: pattern);

        widget = TextSpan(
          text: "${result['display']}",
          style: mapping.style != null ? mapping.style : style,
          recognizer: TapGestureRecognizer()
            ..onTap = () => mapping.onTap!(matchText),
        );
      } else if (mapping.renderWidget != null) {
        widget = WidgetSpan(
          alignment: PlaceholderAlignment.middle,
          child: GestureDetector(
            onTap: () => mapping.onTap!(matchText),
            child: mapping.renderWidget!(
                text: matchText, pattern: mapping.pattern!),
          ),
        );
      } else {
        widget = TextSpan(
          text: "$matchText",
          style: mapping.style != null ? mapping.style : style,
          recognizer: TapGestureRecognizer()
            ..onTap = () => mapping.onTap!(matchText),
        );
      }

      widgets.add(widget);

      return '';
    }, onNonMatch: (String text) {
      widgets.add(TextSpan(
        text: "$text",
        style: this.style,
      ));

      return '';
    });

    if (selectable) {
      return SelectableText.rich(
        TextSpan(children: <InlineSpan>[...widgets], style: style),
        maxLines: maxLines,
        strutStyle: strutStyle,
        textWidthBasis: textWidthBasis,
        textAlign: alignment,
        textDirection: textDirection,
        onTap: onTap as void Function()?,
      );
    }

    return RichText(
      softWrap: softWrap,
      overflow: overflow,
      textScaleFactor: textScaleFactor,
      maxLines: maxLines,
      strutStyle: strutStyle,
      textWidthBasis: textWidthBasis,
      textAlign: alignment,
      textDirection: textDirection,
      text: TextSpan(
        text: '',
        children: <InlineSpan>[...widgets],
        style: style,
      ),
    );
  }
}


lib/src/regex_options.dart
part of flutter_parsed_text;

/// RegexOptions
class RegexOptions {
  /// Creates a RegexOptions object
  /// If `multiLine` is enabled, then `^` and `$` will match the beginning and
  ///  end of a _line_, in addition to matching beginning and end of input,
  ///  respectively.
  ///
  ///  If `caseSensitive` is disabled, then case is ignored.
  ///
  ///  If `unicode` is enabled, then the pattern is treated as a Unicode
  ///  pattern as described by the ECMAScript standard.
  ///
  ///  If `dotAll` is enabled, then the `.` pattern will match _all_ characters,
  ///  including line terminators.
  const RegexOptions(
      {this.multiLine = false,
      this.caseSensitive = true,
      this.unicode = false,
      this.dotAll = false});

  final bool multiLine;
  final bool unicode;
  final bool caseSensitive;
  final bool dotAll;
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_parsed_text/flutter_parsed_text.dart';
import 'package:url_launcher/url_launcher.dart' show canLaunch, launch;

main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'example_app',
      home: MainApp(),
    );
  }
}

class MainApp extends StatefulWidget {
  @override
  _MainAppState createState() => _MainAppState();
}

class _MainAppState extends State<MainApp> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          SizedBox(
            height: 40.0,
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ParsedText(
              selectable: true,
              alignment: TextAlign.start,
              // regexOptions: RegexOptions(
              //   caseSensitive: false,
              //   multiLine: true,
              //   dotAll: false,
              //   unicode: false,
              // ),
              text:
                  "[@michael:51515151] Hello  --- spoiler ---\n\n spoiler content \n--- spoiler ---\n  https://172.0.0.1 london this is https://apps.apple.com/id/app/facebook/id284882215  an example of the ParsedText, links like http://www.google.com or http://www.facebook.com are clickable and phone number 444-555-6666 can call too. But you can also do more with this package, for example Bob will change style and David too.\nAlso a US number example +1-(800)-831-1117. foo@gmail.com And the magic number is 42! #flutter #flutterdev",
              parse: <MatchText>[
                MatchText(
                    type: ParsedType.EMAIL,
                    style: TextStyle(
                      color: Colors.red,
                      fontSize: 24,
                    ),
                    onTap: (url) {
                      launch("mailto:" + url);
                    }),
                MatchText(
                    type: ParsedType.URL,
                    style: TextStyle(
                      color: Colors.blue,
                      fontSize: 24,
                    ),
                    onTap: (url) async {
                      var a = await canLaunch(url);

                      if (a) {
                        launch(url);
                      }
                    }),
                MatchText(
                    type: ParsedType.PHONE,
                    style: TextStyle(
                      color: Colors.red,
                      fontSize: 24,
                    ),
                    onTap: (url) {
                      launch("tel:" + url);
                    }),
                MatchText(
                  type: ParsedType.CUSTOM,
                  pattern:
                      r"^(?:http|https):\/\/[\w\-_]+(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)",
                  style: TextStyle(color: Colors.lime),
                  onTap: (url) => print(url),
                ),
                MatchText(
                    type: ParsedType.CUSTOM,
                    pattern:
                        "(---( )?(`)?spoiler(`)?( )?---)\n\n(.*?)\n( )?(---( )?(`)?spoiler(`)?( )?---)",
                    style: TextStyle(
                      color: Colors.purple,
                      fontSize: 50,
                    ),
                    onTap: (url) {
                      launch("tel:" + url);
                    }),
                MatchText(
                  pattern: r"\[(@[^:]+):([^\]]+)\]",
                  style: TextStyle(
                    color: Colors.green,
                    fontSize: 24,
                  ),
                  renderText: ({required pattern, required str}) {
                    RegExp customRegExp = RegExp(r"\[(@[^:]+):([^\]]+)\]");
                    Match match = customRegExp.firstMatch(str)!;

                    print('test test: ${match[1]}');
                    // return Container(
                    //   padding: EdgeInsets.all(5.0),
                    //   color: Colors.amber,
                    //   child: Text(match[1]!),
                    // );
                    //
                    return {'display': match[1]!};
                  },
                  onTap: (url) {
                    showDialog(
                      context: context,
                      builder: (BuildContext context) {
                        Map<String, String> map = Map<String, String>();
                        RegExp customRegExp = RegExp(r"\[(@[^:]+):([^\]]+)\]");
                        Match match = customRegExp.firstMatch(url)!;
                        // return object of type Dialog
                        return AlertDialog(
                          title: new Text("Mentions clicked"),
                          content: new Text("${match.group(1)!} clicked."),
                          actions: <Widget>[
                            // usually buttons at the bottom of the dialog
                            new FlatButton(
                              child: new Text("Close"),
                              onPressed: () {},
                            ),
                          ],
                        );
                      },
                    );
                  },
                  // onLongTap: (url) {
                  //   print('long press');
                  // },
                ),
                MatchText(
                  pattern: r"\B#+([\w]+)\b",
                  style: TextStyle(
                    color: Colors.pink,
                    fontSize: 24,
                  ),
                  onTap: (url) async {
                    showDialog(
                      context: context,
                      builder: (BuildContext context) {
                        // return object of type Dialog
                        return AlertDialog(
                          title: new Text("Hashtag clicked"),
                          content: new Text("$url clicked."),
                          actions: <Widget>[
                            // usually buttons at the bottom of the dialog
                            new FlatButton(
                              child: new Text("Close"),
                              onPressed: () {},
                            ),
                          ],
                        );
                      },
                    );
                  },
                  // onLongTap: (url) {
                  //   print('long press');
                  // },
                ),
                MatchText(
                    pattern: r"lon",
                    style: TextStyle(
                      color: Colors.pink,
                      fontSize: 24,
                    ),
                    onTap: (url) async {})
              ],
              style: TextStyle(
                fontSize: 24,
                color: Colors.black,
              ),
            ),
          )
        ],
      ),
    );
  }
}
